### PR TITLE
Alternative for instruct mode (Vicuna support, etc.)

### PR DESCRIPTION
**Pull request by aroidzap**
- Added support for `--config` (loading arguments from .txt file).
- Added support for **multiline mode**, disabled by `--disable-multiline` (allows to input newlines - **mode is toggled** by sending `EOF` Ctrl+D on Linux/Mac, Ctrl+Z and Return on Windows)
- Marked `-ins` mode as deprecated. Now it could be replaced with user defined `--ins-prefix`, `--ins-suffix`.
- Added `--stop-prompt` alternative to `--reverse-prompt` - works same, but without adding instruct prefixes and suffixes (useful, when you're talking to agent and you want to force its response to something)
- Added flexibility to `--reverse-prompt` and `--stop-prompt`, to stop, even when trailing space character is missing (because tokens usually starts with space character, and stop/reverse prompt won't trigger otherwise), enabled with `--rm-trailing-space-workaround`
---
There are some code duplicities, some todos and It's overall a bit messy. 
But it should be working and could be merged, if you don't mind.

### CODE REVIEW

File: examples/common.cpp [hunk 1]
```diff
@@ -31,6 +33,43 @@ extern "C" __declspec(dllimport) int __stdcall WideCharToMultiByte(unsigned int
 #define CP_UTF8 65001
 #endif
 
+void split_args(std::string args_string, std::vector<std::string> & output_args)
```
Comment (aroidzap) [inline examples/common.cpp:36]: Could be possibly rewritten with `std::quoted`, but requires C++14.

File: examples/common.cpp [hunk 2]
```diff
@@ -31,6 +33,43 @@ extern "C" __declspec(dllimport) int __stdcall WideCharToMultiByte(unsigned int
...
+        }
+    }
+
+    if (current_arg != "") {
+        output_args.push_back(current_arg);
+    }
+}
+
+std::string unescape(std::string str) {
+    return std::regex_replace(str, std::regex("\\\\n"), "\n");
```
Comment (aroidzap) [inline examples/common.cpp:70]: Also could be implemented better, this is just used when parsing config txt file (arguments).

File: examples/main/main.cpp [hunk 3]
```diff
@@ -298,7 +308,7 @@ int main(int argc, char ** argv) {
             }
 
             // replace end of text token with newline token when in interactive mode
-            if (id == llama_token_eos() && params.interactive && !params.instruct) {
+            if (id == llama_token_eos() && params.interactive && params.input_prefix.empty()) {
```
Comment (aroidzap) [inline examples/main/main.cpp:311]: Not so sure about this change.

File: examples/main/main.cpp [hunk 4]
```diff
@@ -344,27 +354,63 @@ int main(int argc, char ** argv) {
...
+                    if (last_output.find(prompt.c_str(), last_output.length() - antiprompt.len, antiprompt.len) != std::string::npos) {
                         is_interacting = true;
-                        is_antiprompt = true;
+                        antiprompt.any = true;
                         set_console_color(con_st, CONSOLE_COLOR_USER_INPUT);
                         fflush(stdout);
                         break;
                     }
                 }
+                if (!antiprompt.any) {
```
Comment (aroidzap) [inline examples/main/main.cpp:378]: Unnecessary code duplication.

File: examples/main/main.cpp [hunk 5]
```diff
@@ -404,27 +445,40 @@ int main(int argc, char ** argv) {
...
+                }
+
                 // Add tokens to embd only if the input buffer is non-empty
                 // Entering a empty line lets the user pass control back
                 if (buffer.length() > 1) {
 
-                    // instruct mode: insert instruction prefix
-                    if (params.instruct && !is_antiprompt) {
+                    // insert input prefix
+                    if (!params.input_prefix.empty() && !antiprompt.any) {
```
Comment (aroidzap) [inline examples/main/main.cpp:472]: Not so sure about this change.

File: examples/main/main.cpp [hunk 6]
```diff
@@ -441,7 +495,7 @@ int main(int argc, char ** argv) {
 
         // end of text token
         if (!embd.empty() && embd.back() == llama_token_eos()) {
-            if (params.instruct) {
+            if (params.interactive && !params.input_prefix.empty()) {
```
Comment (aroidzap) [inline examples/main/main.cpp:498]: Not so sure about this change.

**Comment by prusnak** [comment]
I like this approach!

One idea, I think it would be nice if configs had only one option per line. This would make future changes more easy to review. Since you already allow empty lines in the config files, we can group options by separating them with empty lines. No need to specify multiple options on one line.

**Comment by anzz1** [comment]
While good as an idea, input prefix/suffix and prompt/reverse prompt are not the same thing and should not be put together or it messes with the tokenization. 

`-r "User:" --in-prefix " "` and `-r "User: "` are different things with different results.

More info about this in the original --in-prefix PR #426

**Comment by aroidzap** [comment]
@anzz1 I have tried to update a bit my implementation to revert changes to input-prefix.

**Comment by aroidzap** [comment]
I have added also "multiline" mode, that does not send input, until EOF is sent (Ctrl+D on Linux/Mac, Ctrl+Z on Windows). I was also trying to look into something like Shift+Enter and Ctrl+V to keep multi-lines (it somehow worked for Windows, but was a bit complicated for other plaftorms). In the end, I have dropped it - it is just not worth hacking CLI terminal to support fancy features. It it would be probably best, if llama.cpp would work as server with http and rest api later (possibly with multiple agents running in parallel as it is possible with mmap approach now).

File: examples/main/main.cpp [hunk 7]
```diff
@@ -25,7 +25,8 @@ static bool is_interacting = false;
 #if defined (__unix__) || (defined (__APPLE__) && defined (__MACH__)) || defined (_WIN32)
 void sigint_handler(int signo) {
     set_console_color(con_st, CONSOLE_COLOR_DEFAULT);
-    printf("\n"); // this also force flush stdout.
+    fflush(stdout);
+    fflush(stderr);
```
Comment (aroidzap) [inline examples/main/main.cpp:29]: Maybe not needed, but it seemed better. Before I had sometimes colored terminal after killing llama.cpp. (Not sure if that is fixed by this though)

File: examples/common.cpp [hunk 8]
```diff
@@ -438,3 +442,51 @@ void win32_utf8_encode(const std::wstring & wstr, std::string & str) {
...
+        if (std::wcin.eof()) {
+            another_line = false;
+            std::wcin.clear();
+            std::wcin.seekg(0, std::ios::beg);
+        }
+        win32_utf8_encode(wline, line);
+#else
+        if (!std::getline(std::cin, line)) {
+            // input stream is bad or EOF received
+            if (std::wcin.bad()) {
```
Comment (j-f1) [inline examples/common.cpp:477]: maybe?
```suggestion
            if (std::cin.bad()) {
```

File: examples/main/main.cpp [hunk 9]
```diff
@@ -147,22 +150,20 @@ int main(int argc, char ** argv) {
     }
 
     // number of tokens to keep when resetting context
-    if (params.n_keep < 0 || params.n_keep > (int)embd_inp.size() || params.instruct) {
+    if (params.n_keep < 0 || params.n_keep > (int)embd_inp.size()) {
```
Comment (aroidzap) [inline examples/main/main.cpp:153]: FYI, user now should explicitly call `--keep -1`. This is added to configs, so it should not obstacle for new users.

File: examples/main/main.cpp [hunk 10]
```diff
@@ -227,8 +228,18 @@ int main(int argc, char ** argv) {
 #if defined (__unix__) || (defined (__APPLE__) && defined (__MACH__)) || defined (_WIN32)
                " - Press Ctrl+C to interject at any time.\n"
 #endif
-               " - Press Return to return control to LLaMa.\n"
-               " - If you want to submit another line, end your input in '\\'.\n\n");
+        );
+        if (params.multiline_mode) {
+#if defined (_WIN32)
+            fprintf(stderr, " - Press Ctrl+Z and Return (EOF) to return control to LLaMa.\n\n");
```
Comment (j-f1) [inline examples/main/main.cpp]:
```suggestion
            fprintf(stderr, " - Press Ctrl+Z then Return (EOF) to return control to LLaMa.\n\n");
```

Comment (aroidzap) [reply examples/main/main.cpp]: Already fixed my typos, thanks.

File: examples/common.cpp (hunk 8 above)
Comment (aroidzap) [reply examples/common.cpp:477]: Already fixed my typos, thanks.

**Comment by aroidzap** [comment]
Changed multiline mode to toggled by sending `EOF`.

File: examples/common.cpp (hunk 1 above)
Comment (Fabio3rs) [reply examples/common.cpp:37]: I think should use const std::string &args_string

File: examples/common.cpp (hunk 2 above)
Comment (Fabio3rs) [reply examples/common.cpp:71]: Here I think should use const std::string &str too

**Comment by slaren** [comment]
Should we split main.cpp into two examples, one basic generation example that only takes a prompt and lets the LLM generate uninterrupted, and one chat example with all of this? I am afraid that main.cpp is becoming too complex and it is not really a good example for getting started with the library anymore.

**Comment by prusnak** [comment]
> Should we split main.cpp into two examples

I think this is a great idea, but I'd probably save it after this PR is merged.

**Comment by niltonvasques** [comment]
It is really making the tests more easy with the new `--config` option. I tested it here and it working fine.

**Comment by ivanstepanovftw** [comment]
Why is there reinvention of wheel if we already can define arguments in both shell and PowerShell like:
```shell
a=(
"Many   spaces" $'And\nnewlines'
1337 -42
)
echo "${a[@]}"
```
```
Many   spaces And
newlines 1337 -42
```

```ps
$arguments = @(
    "Many   spaces",
    "And`nnewlines",
    1337,
    -42
)
.\main @arguments
```

**Review by ggerganov — APPROVED** [review]
`main.cpp` could use some refactoring like moving small parts of the code in utility functions.
Could also make a separate `minimal` example just for text generation based on given prompt. But this can be done later

**Comment by aroidzap** [comment]
Can I merge it now? Or is there something other I should address?

**Comment by rabidcopy** [comment]
I'm not sure what is going on but I can't seem to use basic interactive mode with a reverse prompt anymore. It refuses to give any output and continuously hands back control to me. Disabling multiline mode doesn't seem to change this. I'm thoroughly confused.

**Comment by aroidzap** [comment]
That's weird, we could probably revert this MR until it's solved.

**Comment by ggerganov** [comment]
If needed, it's OK to revert. I won't be able to look at the examples code soon, so fully counting on the collaborators to assist with this

**Comment by prusnak** [comment]
Revert in https://github.com/ggerganov/llama.cpp/pull/982

@aroidzap please open a new pull request when you are ready

**Comment by KevinColemanInc** [comment]
Could we break this MR into smaller ones for each feature? A bug with one feature is preventing all the other features from landing.

**Comment by deadprogram** [comment]
That is interesting, the following command worked well for me from the `master` branch before this was reverted:

```
./main -m ~/Downloads/alpaca2/ggml-model-q4_1.bin --config ./configs/alpaca.txt --ctx_size 2048 -n -1 -b 256 --top_k 10000 --temp 0.2 --repeat_penalty 1 -t 7 -p "A continuación hay una instrucción que describe una tarea. Escriba una respuesta que complete adecuadamente la solicitud."
```

I did modify `config/alpaca.txt` by changing `---reverse-prompt` to `--stop-prompt`

```
$ git diff configs/alpaca.txt
diff --git a/configs/alpaca.txt b/configs/alpaca.txt
index 99a3ab4..3a65e61 100644
--- a/configs/alpaca.txt
+++ b/configs/alpaca.txt
@@ -4,6 +4,6 @@
 --ins-prefix-bos
 --ins-prefix "\n\n### Instruction:\n\n"
 --ins-suffix "\n\n### Response:\n\n"
---reverse-prompt "### Instruction:\n\n"
+--stop-prompt "### Instruction:\n\n"
```

Perhaps a session with `git bisect` could help? @rabidcopy which model did you use when you ran into this issue?

**Comment by rabidcopy** [comment]
> Perhaps a session with `git bisect` could help? @rabidcopy which model did you use when you ran into this issue?

All of them. Vicuna, Vicuna 1.1, Alpaca, gpt4all, gpt4-x-alpaca, etc. They all dole out a token per line and give back control.

However I think I'm on to something. When using the `--config` option instead of manually setting options in the terminal. The new instruct mode works. The only difference being that `--rm-trailing-space-workaround` was passed this time from `vicuna.txt`. I put `--rm-trailing-space-workaround` into my terminal command that previously had issues. Now the issue is gone and things seem to work. New observed issue, pressing Ctrl+D to toggle multi-line mode only works to enable it, and can't seem to be disabled after that. (thus the session is effectively over because you will only be able to input new lines and not get output)

**Comment by jxy** [comment]
It looks like just a few really wrong shell scripts, in which the author falsely think a shell string `"\n"` represent a single newline, `0x0a`.  Typically it's the shell builtin `echo` does the translation of `\`. So just change those to something like

```
    --ins-prefix '

### Instruction:

'
```

**Pull request merged**